### ISSUE TITLE
Implement a tutorial for partial SVDs and partial eigenvalue decompositions

### ISSUE THREAD

**Issue by pnkraemer** [pnkraemer/matfree#224]
At the moment, there is no tutorial for using `matfree.eig`, and since in #221, there were many (very reasonable!) questions about this module, we might benefit from a proper tutorial.

**Comment by pdcook** [comment]
Thanks for such a quick response! Unfortunately, this example behaves the same as the one I had finally got working for myself. The eigenvalues returned don't appear anywhere in the exact dense eigendecomposition.

```python3
import jax.random
import jax.numpy as jnp
from matfree import decomp, eig

M = jax.random.normal(jax.random.PRNGKey(1), shape=(10, 10))
A = M + M.T
v = jax.random.normal(jax.random.PRNGKey(2), shape=(10,))

# Replace tridiagonalisation with bidiagonalisation and eigh with svd
# to compute a partial SVD instead of a partial eigendecomposition.
tridiag = decomp.tridiag_sym(4)
eigh_fun = eig.eigh_partial(tridiag)
eigvals, eigvecs = eigh_fun(lambda s: A @ s, v)

E, V = jnp.linalg.eigh(A)
print("Eigenvalues from matfree eigendecomposition:")
print(eigvals)
print("Eigenvalues from dense eigendecomposition:")
print(E)

"""
Eigenvalues from matfree eigendecomposition:
[-6.815381  -2.0068727  3.6251535  7.2477403]
Eigenvalues from dense eigendecomposition:
[-7.734001  -5.6636744 -4.0365124 -1.9628791 -1.0701044  2.198716
  3.052734   4.061115   4.917902   7.3098483]
"""
```

**Comment by pnkraemer** [comment]
Thanks for following up! What you're seeing is expected behaviour: the eigenvalue computations are approximate, which means that the partial eigenvalues are not guaranteed to be a subset of the full eigendecomposition. Instead, the eigh-via-Lanczos algorithm gravitates towards larger eigenvalues. In the example above, the 7.3 eigenvalue is approximately covered.


In general, this way of partial eigendecompositions works best for matrices whose eigenvalues decay aggressively (think: almost low rank). The random matrix in the example is not one of those because it does not contain eigenvalues that are significantly larger than the others. In other words, the random matrix is a relatively difficult problem for partial eigendecompositions to get right. 

For a more favourable example, replace the matrix A with a matrix whose spectrum decays quickly, e.g. a Gaussian process covariance matrix:

```python
M = jax.random.normal(jax.random.PRNGKey(1), shape=(10,))
A = jnp.exp(-(M[:, None] - M[None, :]) ** 2)
v = jax.random.normal(jax.random.PRNGKey(2), shape=(10,))

# Replace tridiagonalisation with bidiagonalisation and eigh with svd
# to compute a partial SVD instead of a partial eigendecomposition.
tridiag = decomp.tridiag_sym(7)
eigh_fun = eigh_partial(tridiag)
eigvals, eigvecs = eigh_fun(lambda s: A @ s, v)

print(eigvals)
# [2.2047782e-06 2.7635861e-03 2.3097042e-02 3.0454072e-01 8.4691942e-01 2.1064670e+00 6.7161245e+00]

print(jnp.linalg.eigvalsh(A))
# [-1.2856923e-08  1.6782994e-07  4.2913854e-07  8.1828694e-05 2.7639463e-03  2.3100846e-02  3.0454078e-01  8.4691960e-01 2.1064672e+00 6.7161255e+00]
```

In this example, all but the smallest estimated eigenvalue are estimated accurately, and the smallest one is close to machine precision, so it is fair that the algorithm struggles to get that one.

Does that help?

**Comment by pdcook** [comment]
I see. I was approaching this more like `jax.experimental.sparse.lobpcg_standard` or `scipy.sparse.linalg.eigsh` where the number of eigenpairs and the accuracy of those eigenpairs are decoupled. I see that with `matfree` I can get the eigenvalues exactly if I use `num_matvecs` equal to the size of the matrix.

In my use case, the `matvec` function is parameterized and specific parameter sets can have degenerate or near-degenerate extremal eigenvalues. So I think that means `matfree` won't work for my use?

**Comment by pnkraemer** [comment]
Yes, the number of eigenpairs and their accuracy are somewhat coupled. There are extensions to resolve this limitation, eg via restarted Lanczos/Arnoldi, but these more sophisticated algorithms are not implemented currently. (Happy to discuss Matfree implementations though, see #125).

For what it's worth, I remember seeing someone recommend choosing the Krylov-space depth (no. matvecs) to be ~1.5 times the number of desired eigenvalues, and I found this quite reliable. Unfortunately, I don't remember the source. Still, it might be worth trying? Whether or not this is good enough for your application, I can't say, unfortunately.
